[WTH-35] Weeth 서버 성능 모니터링 추가

jj0526 opened the pull request:
## PR 내용
- 프로메테우스 & 그라파나 모니터링 추가

<br>

## PR 세부사항
- `/actuator/prometheus`에서 정보를 가져올 때마다 `error token: 헤더에서 토큰을 찾을 수 없습니다.` 경고 메세지가 뜨는 문제가 발생
  - 회의 결과 토큰이 아예 없는 경우엔 로그를 남기지 않도록 수정
- docker-compose 수정
  - prometheus, grafana, node-exporter 서비스 추가
  - 현재 dev 서버에는 적용하지 않았으며, 머지 시점에 적용 예정

**추가**
-  IP 화이트리스트를 통해 ec2 및 도커 컨테이너 내부 인터넷만 `/actuator/prometheus` 접근 가능하도록 수정
- `/actuator/health` 외부 접근 허용
- docker-compose.yml 에서 프로메테우스, node-exporter 포트는 내부 컨테이너만 접근 가능하도록 수정

<br>

## 관련 스크린샷
- 서버 전체 CPU 예시
<img width="2780" height="1566" alt="image" src="https://github.com/user-attachments/assets/89b52858-ef83-479f-ac89-b0b736c77301" />
- 메트릭
<img width="3434" height="1704" alt="image" src="https://github.com/user-attachments/assets/58c465cb-8f0b-40dd-be92-183649f04ac2" />


<br>

## 주의사항
- ~~Node Exporter, Prometheus, Grafana 포트는 외부 인터넷에서 직접 접근하지 못하도록 보안 그룹에서 제한해야합니다~~
- 컨테이너 내부에서만 접근 가능하도록 수정
<br>

## 체크 리스트

- [x] 리뷰어 설정
- [x] Assignee 설정
- [ ] Label 설정
- [x] 제목 양식 맞췄나요? (ex. #0 Feat: 기능 추가)
- [x] 변경 사항에 대한 테스트

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

## 릴리스 노트

* **New Features**
  * 애플리케이션 모니터링용 Prometheus 통합 추가로 메트릭 수집 및 조회 지원

* **Bug Fixes**
  * 인증 토큰 미발견 상황에 대한 예외 처리 및 디버그 로그 추가로 인증 흐름 안정성 향상

* **Chores**
  * 관리 엔드포인트 노출 범위 조정: 헬스는 공개 허용, Prometheus는 내부 IP에서만 접근 허용 및 메트릭 내보내기 활성화
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jj0526 committed:
chore: 프로메테우스 빌드 및 설정 추가

jj0526 committed:
chore: /actuator/prometheus 접근 인증 제외

jj0526 committed:
feat: TokenNotFoundException 발생 시 로그 생략 추가

jj0526 committed:
feat: TokenNotFoundException 발생 시 debug 로그 추가

jj0526 committed:
chore: actuator 노출 endpoint를 health,prometheus로 제한

hyxklee reviewed (commented):
고생하셨습니다!
소스 코드만으로는 검증이 어려운 부분이 있어서 노션 BE 파트에 도커 컴포즈 설정 파일도 함께 올리고 슬랙으로 공유 부탁드릴게요!

seokjun01 commented on src/main/java/leets/weeth/global/config/SecurityConfig.java:
```diff
@@ -80,6 +80,7 @@ public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
                                                 "/js/**", "/img/**", "/scss/**", "/vendor/**").permitAll()
                                         // 스웨거 경로
                                         .requestMatchers("/v3/api-docs", "/v3/api-docs/**", "/swagger-ui.html", "/swagger-ui/**", "/swagger/**").permitAll()
+                                        .requestMatchers("/actuator/prometheus").permitAll()
```
> permitAll을 해두면 해당 경로가 public 하게 열려서 외부에서도 데이터를 퍼갈 수 있을 것으로 보이는데 별도의 인증책이 필요하지 않을까요??

아니면 Spring Security에서는 permitAll이라 하되 , 보안그룹에서 인바운드 규칙으로 설정을 하는 방식은 어떤가요 ?

seokjun01 commented:
고생하셨습니다!
프로메테우스가  메트릭이라고하는 일종의 데이터를 수집해서 Grafana로 시각화하는 원리 맞을까요 ? 
pull을 받아오는 방식이라고 알고 있는데, 일정한 주기마다 pull받게끔 설정할 수 있는지 궁금합니다

jj0526 committed:
chore: /actuator/health 엔드포인트 접근 허용 설정 추가

jj0526 committed:
feat: 접근 네트워크 수정
- EC2 내부 또는 같은 컨테이너 내에서만 /actuator/prometheus 접근 허용
- /actuator/health 엔드포인트 외부 접근 허용

jj0526 commented on src/main/java/leets/weeth/global/auth/jwt/filter/JwtAuthenticationProcessingFilter.java:
```diff
@@ -47,6 +47,8 @@ protected void doFilterInternal(HttpServletRequest request, HttpServletResponse
             if (jwtProvider.validate(accessToken)) {
                 saveAuthentication(accessToken);
             }
+        } catch (TokenNotFoundException e) {
```
여러 방식을 생각해본 결과 JWT 토큰을 인한 인증은 토큰 만료 관리가 어렵다는 점, 그리고 필터 추가는 향후 유지보수가 어렵고 복잡해진다는 점, 
스프링 시큐리티 설계 원칙 중에서 `인증 로직은 인가 판단에 관여하지 않는다`가 있어, JWT 필터는 인증이 없으면 그냥 지나쳐야 하기 때문에, 토큰이 없을 경우의 로그는 남기지 않는게 좋다고 판단되어 debug만 남기고
토큰이 없을 경우 로그가 남지 않고 IP 화이트리스트를 추가하였습니다

jj0526 committed:
chore: management endpoints 설정을 리스트 형식으로 변경
